File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public int minDays(int[] nums, int m, int k) {
        int n = nums.length;
        if(m*k > n) return -1;
        int max = nums[0];
        int min = max;
        int ans = 0;
        for(int i = 0; i < n; i++) {
            if(nums[i] > max) max = nums[i];
            if(nums[i] < min) min = nums[i];
        }  
        int pairs = m, mid = -1;
        while(min <= max) {
            mid = min + (max - min)/2;
            if(count(nums, mid, m, k)) {
                max = mid - 1;
            } else min = mid + 1;

        }
            return min;

    }
    static boolean count(int[] nums, int day, int m , int k) {
        int n = nums.length;
        int ct = 0;
        int pairs = 0;

        for(int i = 0; i < n; i++) {
            if(nums[i] <= day) {
                ct++;
            }
            else {
                pairs += ct/k;
                ct = 0;
            }
        }
        pairs += ct/k;
        return (pairs >=m);
    }
}

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.py
class Solution:
    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:
        
        start = min(bloomDay)
        end = max(bloomDay)
        minV = end
        if m*k > len(bloomDay):
            return - 1
        while start <= end:
            count = 0
            x = 0
            mid = start + (end - start) // 2
            for num in bloomDay:
                if num <= mid:
                    count += 1
                    if count % k == 0:
                        x += 1
                else:
                    count = 0
            if x >= m:
                end = mid - 1
                if m <= x:
                    minV = min(minV, mid)
            else:
                start = mid + 1
        
        return minV